flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/expressions/fieldExpression.scala

```diff
@@ -103,71 +103,6 @@ case class TableReference(name: String, tableOperation: QueryOperation)
   override def toString: String = s"$name"
 }
 
-abstract class TimeAttribute(val expression: PlannerExpression)
-  extends UnaryExpression
-  with WindowProperty {
-
-  override private[flink] def child: PlannerExpression = expression
-}
-
-case class RowtimeAttribute(expr: PlannerExpression) extends TimeAttribute(expr) {
-
-  override private[flink] def validateInput(): ValidationResult = {
-    child match {
-      case WindowReference(_, Some(tpe: TypeInformation[_])) if isProctimeIndicatorType(tpe) =>
-        ValidationFailure("A proctime window cannot provide a rowtime attribute.")
-      case WindowReference(_, Some(tpe: TypeInformation[_])) if isRowtimeIndicatorType(tpe) =>
-        // rowtime window
-        ValidationSuccess
-      case WindowReference(_, Some(tpe)) if tpe == Types.LONG || tpe == Types.SQL_TIMESTAMP =>
-        // batch time window
-        ValidationSuccess
-      case WindowReference(_, _) =>
-        ValidationFailure("Reference to a rowtime or proctime window required.")
-      case any =>
-        ValidationFailure(
-          s"The '.rowtime' expression can only be used for table definitions and windows, " +
-            s"while [$any] was found.")
-    }
-  }
-
-  override def resultType: TypeInformation[_] = {
-    child match {
-      case WindowReference(_, Some(tpe: TypeInformation[_])) if isRowtimeIndicatorType(tpe) =>
-        // rowtime window
-        TimeIndicatorTypeInfo.ROWTIME_INDICATOR
-      case WindowReference(_, Some(tpe)) if tpe == Types.LONG || tpe == Types.SQL_TIMESTAMP =>
-        // batch time window
-        Types.SQL_TIMESTAMP
```

Comment:
Does this test need to return an SQL_TIMESTAMP:

```
                TestSpec.forStrategy(
                                "ROWTIME type strategy on long in batch mode",
                                SpecificInputTypeStrategies.windowTimeIndicator(
                                        TimestampKind.ROWTIME))
                        .calledWithArgumentTypes(DataTypes.BIGINT())
                        .expectArgumentTypes(DataTypes.BIGINT()));
                        ```

Reply:
The test you posted covers the input strategy not output. However you're right, it should return `TIMESTAMP(3)` (SQL_TIMESTAMP is a legacy type). I updated the output strategy for `ROWTIME`

Reply:
Just to check my understanding, we do not need to update the output strategy for proctime since its type is constant, right?

Reply:
Actually we do need to update it for `PROCTIME`, because we always need to return `PROCTIME time indicator`.

I thought having `timestampKind == TimestampKind.PROCTIME && !LogicalTypeChecks.isTimeAttribute(type)` for input is enough, but it passes also for `ROWTIME time indicator`.

I'll update the output type strategy for `PROCTIME`

Reply:
Nice!  Glad I asked, and good job finding the corner/edge case.